### PR TITLE
Add test for `write.metadata.compression-codec` property in Iceberg

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Add test for `write.metadata.compression-codec` property in Iceberg

Property | Default | Description
-- | -- | --
write.metadata.compression-codec | none | Metadata compression codec; none or gzip

https://iceberg.apache.org/docs/latest/configuration/

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java [hunk 1]
```diff
@@ -60,6 +61,12 @@ public IcebergTableProperties(
                         IcebergFileFormat.class,
                         icebergConfig.getFileFormat(),
                         false))
+                .add(enumProperty(
+                        METADATA_COMPRESSION_CODEC,
+                        "Metadata compression codec; none or gzip",
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:66]: We don't need to list the possible values in the description. This description may become obsolete when introducing a different compression codec.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java [hunk 2]
```diff
@@ -2087,6 +2087,38 @@ public void testHandlingPartitionSchemaEvolutionInPartitionMetadata()
...
+        String trinoTableName = trinoTableName(baseTableName);
+        String sparkTableName = sparkTableName(baseTableName);
+
+        onSpark().executeQuery("CREATE TABLE " + sparkTableName + "(col int) USING iceberg TBLPROPERTIES ('write.metadata.compression-codec'='gzip')");
+        onSpark().executeQuery("INSERT INTO " + sparkTableName + " VALUES (1)");
+
+        assertThat(onTrino().executeQuery("SELECT * FROM " + trinoTableName)).containsOnly(row(1));
+
+        // Change 'write.metadata.compression-codec' to none and select the table in Trino
+        onSpark().executeQuery("ALTER TABLE " + sparkTableName + " SET TBLPROPERTIES ('write.metadata.compression-codec'='none')");
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:2118]: We should probably add corresponding logic to `io.trino.plugin.iceberg.IcebergMetadata#setTableProperties` in order for this statement to be effective.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            case NONE -> Codec.NONE;
+            case GZIP -> Codec.GZIP;
+        };
+    }
+
+    public static IcebergMetadataCompressionCodec fromIceberg(Codec format)
+    {
+        return switch (format) {
+            case NONE -> NONE;
+            case GZIP -> GZIP;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:36]: should we add here a `default` option , just in case some new `Codec` value gets added in the following Iceberg versions?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java [hunk 4]
```diff
@@ -111,6 +118,11 @@ public static IcebergFileFormat getFileFormat(Map<String, Object> tablePropertie
         return (IcebergFileFormat) tableProperties.get(FILE_FORMAT_PROPERTY);
     }
 
+    public static IcebergMetadataCompressionCodec getMetadataCompressionCodec(Map<String, Object> tableProperties)
+    {
+        return (IcebergMetadataCompressionCodec) tableProperties.get(METADATA_COMPRESSION_CODEC);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:123]: to be on the safe side - maybe a previous version of Iceberg didn't return this value, should we maybe use `getOrDefault` `IcebergMetadataCompressionCodec.NONE`

**Comment by findinpath** [comment]
```
trino:default> create table test1 (x integer) WITH (metadata_compression_codec='gzip');

trino:default> show create table test1;
                                             Create Table                                             
------------------------------------------------------------------------------------------------------
 CREATE TABLE iceberg.default.test1 (                                                                 
    x integer                                                                                         
 )                                                                                                    
 WITH (                                                                                               
    format = 'PARQUET',                                                                               
    format_version = 2,                                                                               
    location = 'hdfs://hadoop-master:9000/user/hive/warehouse/test1-d1e1ae768e304003b0c322b9a5df5368' 
 )                                                                                                    
(1 row)

```

`metadata_compression_codec` property is missing

**Comment by findinpath** [comment]
Spark Iceberg is able to change on the fly the `write.metadata.compression-codec` property  and this gets reflected in the metadata files created :

```
➜  metadata ls
00000-20dad955-ef84-4b40-b9ae-2cfebcf02d8d.gz.metadata.json
00001-1502c408-3b4d-4358-8265-de5d8e64636e.gz.metadata.json
00002-d2bc1c57-62af-4238-a64c-7d298b53eeca.metadata.json
00003-2c2b5b36-5e37-44ee-8cf6-052fddb92951.metadata.json
```

Notice in the snippet above that there are both `gz.metadata.json` and `metadata.json` files.

Corresponding test scenario:

```
spark-sql> create table test1 (x integer) tblproperties ('write.metadata.compression-codec' = 'gzip');
Response code
Time taken: 2.634 seconds
spark-sql> insert into test1 values 1;
Response code
Time taken: 2.877 seconds
spark-sql> alter table test1 set tblproperties ('write.metadata.compression-codec' = 'none');
Response code
Time taken: 0.075 seconds
spark-sql> insert into test1 values 2;
spark-sql> select * from test1;
x
2
1
Time taken: 0.549 seconds, Fetched 2 row(s)
```


Should we allow changing this property as well in `io.trino.plugin.iceberg.IcebergMetadata#setTableProperties` ?

**Comment by alexjo2144** [comment]
On the testing side, can we assert that the files are actually compressed? Also, there are a few property tests in `TestIcebergV2` like `testUpdatingAllTableProperties` that could be updated

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java (hunk 3 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:36]: It should probably throw if it's not one of these two

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:36]: Though, do we need the Trino version of this enum, or could we just use the Iceberg one?

**Comment by alexjo2144** [comment]
Optional: We could add a`metadata_compression_codec_default` to `IcebergConfig` so if you wanted to have gzip enabled by default you could.

**Comment by ebyhr** [comment]
> metadata_compression_codec property is missing

I intentionally omitted it when the value is none. 

> Should we allow changing this property as well in io.trino.plugin.iceberg.IcebergMetadata#setTableProperties ?

This is also intentionally disallowed to start with small feature (= we can support it when needed).

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:36]: I suppose we can easily miss new values if we directly use Iceberg enum.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 5]
```diff
@@ -319,6 +328,18 @@ public static IcebergFileFormat getFileFormat(Map<String, String> storagePropert
...
 
+    public static IcebergMetadataCompressionCodec getMetadataCompressionCodec(Table table)
+    {
+        return getMetadataCompressionCodec(table.properties());
+    }
+
+    public static IcebergMetadataCompressionCodec getMetadataCompressionCodec(Map<String, String> storageProperties)
+    {
+        return IcebergMetadataCompressionCodec.fromIceberg(Codec.fromName(storageProperties
+                .getOrDefault(METADATA_COMPRESSION_CODEC, METADATA_COMPRESSION_DEFAULT)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:339]: Use 

```
.getOrDefault("write.metadata.compression-codec", METADATA_COMPRESSION_DEFAULT)
```

The storage properties are Iceberg specific, not Trino specific.

Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:339]: Let's modify also `io.trino.plugin.iceberg.BaseIcebergMaterializedViewTest#testShowCreate` to contain a `gzip` metadata compression codec.

optional: We could also add a test for iceberg tables showing the two approaches (when metadata codec is none it doesn't show up in show create, when it is `gzip` it shows up in show create).

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java [hunk 6]
```diff
@@ -2087,6 +2087,38 @@ public void testHandlingPartitionSchemaEvolutionInPartitionMetadata()
...
+        assertThat(onTrino().executeQuery("SELECT * FROM " + trinoTableName)).containsOnly(row(1));
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:2119]: Let's also add an extra insert with Trino :
```
onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (3)");
```

We verify this way that even with mixed metadata codecs, the insertion still works in Trino.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java [hunk 7]
```diff
@@ -2087,6 +2087,38 @@ public void testHandlingPartitionSchemaEvolutionInPartitionMetadata()
...
+    public void testSparkMetadataCompressionCodecGzip()
+    {
+        // Verify that Trino can read a table created by Spark
+        String baseTableName = "test_spark_metadata_compression_codec_gzip" + randomTableSuffix();
+        String trinoTableName = trinoTableName(baseTableName);
+        String sparkTableName = sparkTableName(baseTableName);
+
+        onSpark().executeQuery("CREATE TABLE " + sparkTableName + "(col int) USING iceberg TBLPROPERTIES ('write.metadata.compression-codec'='gzip')");
+        onSpark().executeQuery("INSERT INTO " + sparkTableName + " VALUES (1)");
+
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:2114]: Let's also add an extra insert with Trino :
```
onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (2)");
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 5 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java]: > The storage properties are Iceberg specific, not Trino specific.

Thanks for catching that. It means these constants naming might be little confusing. 

> Let's modify also io.trino.plugin.iceberg.BaseIcebergMaterializedViewTest#testShowCreate to contain a gzip metadata compression codec.

Updated `BaseIcebergConnectorTest#testMetadataCompressionCodecGzip` instead of `BaseIcebergMaterializedViewTest`. 

> We could also add a test for iceberg tables showing the two approaches (when metadata codec is none it doesn't show up in show create, when it is gzip it shows up in show create).

Added `SHOW CREATE TABLE ` check for `gzip` case. Regarding `none` case, it's covered in existing tests `BaseIcebergConnectorTest#testShowCreateTable`.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import org.apache.iceberg.TableMetadataParser.Codec;
+
+public enum IcebergMetadataCompressionCodec
+{
+    NONE,
+    GZIP,
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:21]:
```suggestion
    NONE(Codec.NONE),
    GZIP(Codec.GZIP),
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            default -> throw new IllegalArgumentException("Unsupported Trino metadata compression codec: " + this);
+        };
+    }
+
+    public static IcebergMetadataCompressionCodec fromIceberg(Codec format)
+    {
+        return switch (format) {
+            case NONE -> NONE;
+            case GZIP -> GZIP;
+            default -> throw new IllegalArgumentException("Unsupported Iceberg metadata compression codec: " + format);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:38]: Remove `default`.
It's not necessary today and will force the code to be updated when Iceberg library changes

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java (hunk 8 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:21]: This will change toIceberg() into a property accessor

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java [hunk 10]
```diff
@@ -2087,6 +2087,42 @@ public void testHandlingPartitionSchemaEvolutionInPartitionMetadata()
...
+        // Verify that Trino can read and write to a table created by Spark
+        String baseTableName = "test_spark_metadata_compression_codec_gzip" + randomTableSuffix();
+        String trinoTableName = trinoTableName(baseTableName);
+        String sparkTableName = sparkTableName(baseTableName);
+
+        onSpark().executeQuery("CREATE TABLE " + sparkTableName + "(col int) USING iceberg TBLPROPERTIES ('write.metadata.compression-codec'='gzip')");
+        onSpark().executeQuery("INSERT INTO " + sparkTableName + " VALUES (1)");
+        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (2)");
+
+        assertThat(onTrino().executeQuery("SELECT * FROM " + trinoTableName)).containsOnly(row(1), row(2));
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:2116]: Did this work or fail before?

File: docs/src/main/sphinx/connector/iceberg.rst [hunk 11]
```diff
@@ -565,6 +565,9 @@ Property name                                      Description
                                                    specification to use for new tables; either ``1`` or ``2``.
                                                    Defaults to ``2``. Version ``2`` is required for row level deletes.
 
+``metadata_compression_codec``                     Metadata compression codec;
+                                                   either ``NONE`` or ``GZIP``.  Defaults to ``NONE``.
```
Comment (findepi) [inline docs/src/main/sphinx/connector/iceberg.rst:569]: Under what circumstances a user would want to set this?

cc @danielcweeks @rdblue @alexjo2144

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java (hunk 10 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:2116]: It worked before.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java (hunk 3 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:40]: There is no other default than "throw" and "let it throw" is not what we want. We want to update our enum accordingly.

> Though, do we need the Trino version of this enum, or could we just use the Iceberg one?

Having our own is not a big overhead and lets us use static code analysis to know when to update the docs -- Trino docs enumerate all possible values. If we used Iceberg's class, we would want to defer to Iceberg docs for possible values.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.apache.iceberg.TableMetadataParser.Codec;
+
+public enum IcebergMetadataCompressionCodec
+{
+    NONE(Codec.NONE),
+    GZIP(Codec.GZIP),
+    /**/;
+
+    private final Codec codec;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCompressionCodec.java:24]: i'd probably call it "icebergCodec"

**Review by findepi — APPROVED** [review]

**Comment by rdblue** [comment]
@findepi, I don't think that adding configuration for this is a good idea. This should be configured at the table level, not in processing engines.

**Comment by alexjo2144** [comment]
@rdblue sorry, I'm not sure exactly what you're suggesting. The code here is really just an alias for setting the Iceberg `TableProperties: write.metadata.compression-codec` so the value is being set as a table property, to be used across engines.

**Comment by rdblue** [comment]
@alexjo2144, metadata files are typically handled within Iceberg, so I don't think Trino should need to do anything to make sure `write.metadata.compression-codec` is handled correctly. Trino should not need to worry about this property at all.

**Comment by alexjo2144** [comment]
Trino doesn't have an arbitrary `SET TBLPROPERTIES` syntax that takes any string, property keys need to be explicitly enabled. So, any table properties we want users to be able to enable/disable from Trino need some minimal amount of handling. 

Unless you're saying users shouldn't be able to enable this property from Trino at all?

**Comment by ebyhr** [comment]
> I don't think that adding configuration for this is a good idea. This should be configured at the table level, not in processing engines.

This change doesn't add a configuration. `IcebergTableProperties` is for **table** property.

**Comment by rdblue** [comment]
@ebyhr, @alexjo2144, I didn't realize that there wasn't a way to set table configuration in Trino. I'm still not sure that the right option is to implement PRs like this for individual table properties. And, I'm a bit concerned about creating a custom name that doesn't match Iceberg. That just requires more documentation and fragments support across the ecosystem -- people won't be able to set `metadata_compression_codec` anywhere else. And it's also a bit strange to show the underlying table property.

Can you pass properties from `WITH` through if they aren't mapped to anything specific, like partitioning?

**Comment by findepi** [comment]
> I didn't realize that there wasn't a way to set table configuration in Trino

Trino allows to set table properties that are declared by the connector.

> Can you pass properties from `WITH` through if they aren't mapped to anything specific

Not currently. Trino table properties are strictly typed and so they need to be declared

> like partitioning?

`partitioning` is handled explicitly by Trino


There are two aspects of this

- technical side: how a user should be able to set the `write.metadata.compression-codec` table property
   - exposing as `metadata_compression_codec` is one way of doing it
   - allowing users to set arbitrary string-valued properties is another way of doing this
- usefulness question: Under what circumstances a user would want to set this? (https://github.com/trinodb/trino/pull/13953#discussion_r961581090)
   - if the answer is "never", we should not add a new table property
   - if the answer is "always", we should also not add a table property
   - if the answer is "depends on the metadata size", then I'd argue that maybe we don't need the table property either -- Trino knowns metadata size, so it could apply compression or not automatically.


cc @electrum @dain

**Comment by findepi** [comment]
An administrator/operator is also a user from the Trino project perspective.
When would an administrator want to set/change value of `write.metadata.compression-codec` Iceberg table property?

**Comment by rdblue** [comment]
@findepi, an administrator would set this for all tables in the catalog, not for individual tables.

**Comment by ebyhr** [comment]
@rdblue What's the situation an administrator sets gzip compression for metadata file in the catalog?

**Comment by rdblue** [comment]
@ebyhr, this is a table level setting because that's the granularity at which you can change it. We also don't expect users to interact with this. It is probably carried through from the catalog when a table gets created -- that's what we do. It's reasonable for an engine to allow changing table-level settings like this, but it isn't something that should be specific to an engine (with a special property that only works in Trino).

**Comment by ebyhr** [comment]
@rdblue I hope your will answer "What's the situation an administrator sets gzip compression for metadata file in the catalog?"

**Comment by rdblue** [comment]
@ebyhr, this setting is most likely controlled by administrators that set table defaults. Users will probably never go into this level of detail.

**Comment by ebyhr** [comment]
@rdblue I'm not asking who sets the property. Let me rephrase my question. "When administrators want to use gzip for metadata compression"? e.g. file size > n

**Review by findepi — APPROVED** [review]
